### PR TITLE
keep metadata during pull and FROM

### PR DESCRIPTION
**Pull request by reidpr**
Addresses: #651, #765, ~~#964~~ (last moved to PR #968).
Related: #442.

### CODE REVIEW

**Review by j-ogas — APPROVED** [review]
Super excited about this. I do think the changes that preserve metadata from pull and the limitations, e.g., v2 manifest only, should be reflected in our documentation somewhere. Thoughts?

File: lib/build.py [hunk 1]
```diff
@@ -460,26 +460,26 @@ def dest_realpath(self, unpack_path, dst):
          iter_ct += 1
          if (iter_ct > 100):  # arbitrary
             ch.FATAL("can't COPY: too many path components")
-         ch.DEBUG("current destination: %d %s" % (iter_ct, dst_canon), v=2)
-         #ch.DEBUG("parts remaining: %s" % dst_parts, v=2)
+         ch.TRACE("current destination: %d %s" % (iter_ct, dst_canon))
+         #ch.TRACE("parts remaining: %s" % dst_parts)
```
[the remark this answers is not in the document]
Comment (reidpr) [reply lib/build.py:464]: Yeah, it was to much blabber even for `TRACE`.

**Comment by reidpr** [comment]
> I do think the changes that preserve metadata from pull and the limitations, e.g., v2 manifest only, should be reflected in our documentation somewhere.

I agree. See new commit.

File: lib/charliecloud.py [hunk 2]
```diff
@@ -291,9 +263,98 @@ class defined in this module."""
...
+                        "labels": dict(),
+                        "shell": ["/bin/sh", "-c"],
+                        "volumes": list() }  # set isn't JSON-serializable
+
+   def metadata_load(self):
+      """Load metadata file, replacing the existing metadata object. If
+         metadata doesn't exist, warn and use defaults."""
+      path = self.metadata_path // "metadata.json"
+      if (not path.exists()):
+         WARNING("no metadata for image; using defaults")
```
Comment (j-ogas) [inline lib/charliecloud.py:284]: Perhaps we should specify the image reference in this warning? Instead of just saying "image".

**Review by j-ogas — CHANGES_REQUESTED** [review]

**Pull request merged**